zfs-snapshots.py

```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env python3
+
+import subprocess
+from functools import reduce, partial
+from itertools import groupby
+from operator import itemgetter, add
+from prometheus_client import CollectorRegistry, Gauge, generate_latest
+
+
+def row_to_metric(metric: Gauge, row):
+    return metric.labels(pool=row[0][0], volume=row[0][1]).set(row[1])
+
+
+def collect_metrics(metric: Gauge, it):
+    return list(map(partial(row_to_metric, metric), it))
+
+
+def zfs_parse_line(line):
+    cols = line.split("\t")
+    rest, snapshot = cols[0].rsplit("@", 1)
+    pool = rest
+    volume = None
+    if "/" in rest:
+        pool, volume = rest.split("/", 1)
+        volume = "/" + volume
+    return [pool, volume, snapshot, *map(int, cols[1:])]
+
+
+def zfs_list_snapshots():
+    cmd = [
+        "zfs",
+        "list",
+        "-p",
+        "-H",
+        "-t",
+        "snapshot",
+        "-o",
+        "name,used,creation",
+    ]
+    popen = subprocess.Popen(cmd, stdout=subprocess.PIPE)
```

Comment:
Unless you need to interact with the process while it is running (i.e. send input to its `stdin`), I would be more inclined to simply use `subprocess.run()` with `check=True`.

Reply:
I would also recommend passing `LC_ALL=C` environment variable to the child process, as we often encounter issues where tools output locale-specific numeric formatting, which can trip up the collector (e.g. European-style comma instead of decimal point or vice versa, dots as thousands separator). Something like `env=dict(os.environ, LC_ALL="C"))` passed to `subprocess.run()` should be sufficient.

Reply:
On `subprocess.run()`: I chose popen because I want to process line by line even if the command hasn’t finished yet.

Passing `LC_ALL` makes tons of sense, changed.

Reply:
> On `subprocess.run()`: I chose popen because I want to process line by line even if the command hasn’t finished yet.

Why is that important? The resulting .prom file (e.g. when using `sponge`) is not going to be written until the collector script finishes anyway. Collectors should not output "partially successful" metrics - it should be all or nothing (and exit with an error).

Reply:
It’s not about the output of the collector but about the output of the zfs command. It’s relatively slow (takes on a system with 11K snapshots around 5 seconds) so I am trying to shave off every millisecond I get so processing as soon as any output arrives from the zfs command is to not waste any time in an otherwise already slowish command.

Reply:
Ok, fair enough. Perhaps add a code comment then, explaining the motivation behind it, so that it doesn't merely look like unnecessary over-engineering.

Reply:
Done